### ISSUE TITLE
[chrombpnet-eval] One-hot ChromBPNet training diverges to NaN — gradient clipping, precision, loss range

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/marin-dna#247]
🤖 Sub-issue of #236; surfaced while smoke-testing the M1b one-hot training pipeline (#241).

## Symptom

A second GM12878 one-hot smoke — identical data/config to the first, **plus** `set_float32_matmul_precision('high')` (TF32 matmuls) and the online QTL callback — **diverged to NaN within the first few steps**:

```
train_loss_step:     398  → nan
train_profile_loss:  375  → 334 → nan
train_count_loss:    2.99 → nan
grad_norm:           438  → nan
```

(`val_count_pearson` / `qtl_*_pearson` read `0.000` only because the metric guards map NaN predictions → 0 — the divergence is in **training**, not the metrics.)

W&B: [`agyaot6b`](https://wandb.ai/gonzalobenegas/chrombpnet-eval/runs/agyaot6b) (NaN) vs the earlier stable [`dtzi38a4`](https://wandb.ai/gonzalobenegas/chrombpnet-eval/runs/dtzi38a4).

## Diagnosis

- **Early-training gradient spike.** Both smokes hit a huge spike at step ~3 — full-fp32 smoke #1: `train_loss=3.89e4, train_count_loss=5e3, grad_norm=3.19e5` — then **recovered** and trained to `val_count_pearson≈0.42`. Smoke #2 (TF32 on) hit the same spike but went **NaN** instead. The spike originates in the **count loss** (`MSE(pred_logcount, true_logcount) × alpha`, `alpha = median_count/10 ≈ 7.7`): at random init the predicted log-counts are far off → MSE in the thousands → gradient ~1e5.
- **TF32 likely tips it over.** Full fp32 (smoke #1) survived the spike; reduced-mantissa TF32 matmul (smoke #2) probably overflowed it to inf→NaN. No fixed seed yet, so run-to-run variance can't be *fully* ruled out — but TF32 is the obvious difference.
- **The large profile-loss number (~375) is mostly cosmetic.** The vendored `multinomial_nll` includes the multinomial-coefficient (`lgamma`) term, which is **constant w.r.t. the logits → zero gradient**. So the "crazy range" of the profile loss is a gradient-free offset, not the instability driver — the count-loss spike is.

## Proposed fixes (agree before implementing)

1. **Full precision for now** — default `--matmul-precision highest` (disable the TF32 I added); keep fp32; defer bf16 until training is stable.
2. **Gradient-norm clipping** (`Trainer(gradient_clip_val=…)`) — the direct fix for the early spike. The `grad_norm` we log is pre-clip, so it stays diagnostic. Value TBD (global-norm; pick from the observed pre-clip norms).
3. **LR warmup** — ease the first steps so the count head doesn't slam.
4. **`seed_everything`** — make "stable vs NaN" deterministic / reproducible.
5. **Loss-range (optional, monitoring only)** — drop the gradient-free `lgamma` term for a cleaner profile-loss number; sanity-check `alpha`.

## Open questions

- Clip value — global-norm `1.0`, or scaled to the loss magnitude? Warmup length?
- Does official ChromBPNet / ARSENAL clip or warm up? (worth checking their configs)
- Keep the `lgamma` term (faithful to the vendored loss) or drop it (cleaner number, identical gradients)?

**Comment by gonzalobenegas** [comment]
🤖 The NaN divergence is fixed by #249 (merged) — the PR said it "addresses" this issue but didn't use a closing keyword, so it stayed open. All four substantive fixes proposed here shipped in [`train_onehot.py`](https://github.com/Open-Athena/marin-dna/blob/d1a5b315e052ca78a7c331ec2673c35e091f6116/scripts/chrombpnet_eval/train_onehot.py):

1. **Full-precision default** — [`--matmul-precision highest`](https://github.com/Open-Athena/marin-dna/blob/d1a5b315e052ca78a7c331ec2673c35e091f6116/scripts/chrombpnet_eval/train_onehot.py#L103); the TF32 `high` path that contributed to the divergence is now opt-in only.
2. **Global-norm gradient clipping** — [`--grad-clip`](https://github.com/Open-Athena/marin-dna/blob/d1a5b315e052ca78a7c331ec2673c35e091f6116/scripts/chrombpnet_eval/train_onehot.py#L89) (default 1000), wired to `Trainer(gradient_clip_val=…)`; the logged `grad_norm` stays pre-clip/diagnostic.
3. **LR warmup** — [`--warmup-steps`](https://github.com/Open-Athena/marin-dna/blob/d1a5b315e052ca78a7c331ec2673c35e091f6116/scripts/chrombpnet_eval/train_onehot.py#L83) (default 100), a linear ramp in `ChromBPNetLit`.
4. **`seed_everything`** — [`L.seed_everything(args.seed, workers=True)`](https://github.com/Open-Athena/marin-dna/blob/d1a5b315e052ca78a7c331ec2673c35e091f6116/scripts/chrombpnet_eval/train_onehot.py#L158).

bf16-mixed was then validated to train clean (the fp32-safe count-combine forward), and the full early-stopped GM12878 run reached epoch 12 with 0 NaN.

The **only** leftover is the optional fix #5 — drop the gradient-free `lgamma` multinomial-coefficient term from the profile-loss *number*. It is constant w.r.t. the logits (zero gradient), so it has no training effect; it is purely a cosmetic loss-range readout, and the faithful vendored `multinomial_nll` keeps it. Not worth its own tracking issue — closing as completed. If the readout ever matters it's a one-line change in `_vendor/chrombpnet/model_wrappers.py`.
